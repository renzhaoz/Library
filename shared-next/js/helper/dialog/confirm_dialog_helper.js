/* global LazyLoader */
/* eslint-disable no-unused-vars, no-undef, no-unused-expressions, no-empty-function*/
(function confirmDialogHelper() {
  const resources = [
    '%SHARED_APP_ORIGIN%/elements/gaia_confirm/gaia_confirm.js'
  ];
  if (!window.ComponentUtils) {
    resources.push(
      '%SHARED_APP_ORIGIN%/js/utils/components/component_utils.js'
    );
  }

  let dialogHelperElement = null;
  const confirmSoftkey = new SoftkeyPanel({
    menuClassName: 'menu-button',
    items: [
      {
        name: '',
        priority: 1,
        method() {
          console.log('confirmSoftkey created');
        }
      }
    ]
  });
  confirmSoftkey.hide();

  /**
   * Generic dialog helper this _depends_ on <gaia-confirm> but is not required
   * to be loaded as part of the main gaia-confirm script.
   */
  function ConfirmDialogHelper(config) {
    this.config = config;
  }

  ConfirmDialogHelper.prototype = {
    activeElement: null,
    deltaTop: 100,
    Softkey: confirmSoftkey,
    isShown: false,
    show(parent) {
      LazyLoader.load(resources, this._show.bind(this, parent));
    },

    _show(parent) {
      const { config } = this;
      const self = this;

      if (NavigationMap && NavigationMap.currentActivatedLength > -1) {
        if (!NavigationMap.currentActivatedLength) {
          NavigationMap.currentActivatedLength = 1;
        } else {
          setTimeout(() => {
            self.show(parent);
          }, 600);
          return;
        }
      }
      self.parent = parent;
      self.activeElement = document.activeElement;
      const wrapper = document.createElement('div');

      // Only include a cancel button if the config was given...
      const cancelButton = config.cancel
        ? '<button class="cancel" type="button"></button>'
        : '';
      const confirmClass = config.system ? 'insystem' : '';
      const extraClass = config.extraClass ? config.extraClass : '';
      // This is a hack set role="menuitem" in order to readout all components content.
      const contentStyle =
        'font-weight: normal;' +
        'word-wrap: break-word;' +
        'margin: 0;' +
        'padding: 1rem;' +
        'line-height: 2.4rem;' +
        'color: var(--color-gs90);' +
        'background: var(--color-gs10);';
      wrapper.innerHTML =
        `<gaia-confirm role="menuitem" tabindex="-1" class=${confirmClass} ${extraClass}>` +
        `<h1 slot="title" style="text-align:center;"></h1>` +
        `<div slot="content" class="content"><div>` +
        `<p style="${contentStyle}"></p>` +
        `<p class="noborder" style="${contentStyle}" hidden></p>` +
        `</div></div>` +
        `</gaia-confirm>`;

      const element = wrapper.firstElementChild;

      this.element = element;
      dialogHelperElement = element;

      element.dataset.type = config.type;
      element.addEventListener('confirm', this);
      element.addEventListener('cancel', this);
      element.addEventListener('accept', this);
      window.addEventListener('keydown', this);

      /*
       * XXX: Primarily here for pressing the home screen button.
       * The home button triggers a hashchange of the homescreen.
       */
      window.addEventListener('hashchange', this);
      // TO DO: Add visibility change handling...

      const plist = element.querySelectorAll('p');
      const title = element.querySelector('h1');
      const [body, desc] = plist;

      /*
       *Var cancel = element.querySelector('.cancel');
       *var confirm = element.querySelector('.confirm');
       */

      const setL10nAttributes = function setL10nAttributes(ele, options) {
        if ('string' === typeof options) {
          window.api.l10n.setAttributes(ele, options);
        }

        if (options.id) {
          window.api.l10n.setAttributes(ele, options.id, options.args);
        } else if (options.text) {
          ele.textContent = options.text;
        }
      };

      setL10nAttributes(title, config.title);
      setL10nAttributes(body, config.body);

      if (config.body.html) {
        body.innerHTML = config.body.html;
      }

      if (!config.title.id && !config.title.text) {
        element.querySelector('h1').hidden = true;
      }

      if (config.desc || (config.desc && config.desc !== '')) {
        desc.hidden = false;
        setL10nAttributes(desc, config.desc);
      } else {
        desc.hidden = true;
      }

      element.setAttribute('hidden', '');
      parent.appendChild(element);
      self.activeElement.blur();

      /*
       * This nested requestAnimationFrame is to work around the coalescing
       * of the style changes associated with removing of the 'hidden'
       * attribute with the creation of the element.
       * For whatever reason, flushing the style with the usual trick of
       * accessing clientTop doesn't work, and a setTimeout requires an
       * unreasonably lengthy timeout (>50ms) to work, and that may not be
       * a reliable solution.
       * This work-around, though gross, appears to work consistently without
       * introducing too much lag or extra work.
       */
      window.requestAnimationFrame(() => {
        window.requestAnimationFrame(() => {
          element.removeAttribute('hidden');
          // Call scrollContext here to get correct maxHeight of the content.
          self.scrollContext('ArrowUp');
          self.element.addEventListener('transitionend', function udInput() {
            if (self.input) {
              const len = self.input.value.length;
              self.input.focus();
              self.input.setSelectionRange(len, len);
              self.input.oninput = () => {
                self.updateSoftkey();
              };
            }
            self.element.removeEventListener('transitionend', udInput);
            self.isShown = true;
          });
          window.dispatchEvent(new CustomEvent('gaia-confirm-open'));
        });
      });

      // Init softkey
      this.input = element.querySelector('input');
      this.updateSoftkey();
    },

    updateSoftkey() {
      const softkeyParams = {
        menuClassName: 'menu-button',
        header: {},
        items: []
      };
      const bShowRSK = !this.input || this.input.value.trim().length > 0;
      if (this.config.cancel) {
        softkeyParams.items.push({
          name: '',
          priority: 1,
          l10nId: this.config.cancel.l10nId,
          method() {}
        });
      }
      if (this.config.accept) {
        softkeyParams.items.push({
          name: '',
          priority: 2,
          l10nId: this.config.accept.l10nId,
          icon: this.config.accept.icon || null,
          method() {}
        });
      }
      if (this.config.confirm && bShowRSK) {
        softkeyParams.items.push({
          name: '',
          priority: 3,
          l10nId: this.config.confirm.l10nId,
          method() {}
        });
      }
      confirmSoftkey.initSoftKeyPanel(softkeyParams);
      confirmSoftkey.show();
    },

    close() {
      this.destroy();
      confirmSoftkey.hide();
    },

    destroy() {
      if (!this.element || !this.isShown) {
        return;
      }
      if (NavigationMap && NavigationMap.currentActivatedLength > -1) {
        NavigationMap.currentActivatedLength = 0;
      }
      this.activeElement.focus();
      // Ensure cleanup of our hacks!
      window.removeEventListener('hashchange', this);
      window.removeEventListener('keydown', this);

      this.element.addEventListener('transitionend', e => {
        if (e.target !== this.element) {
          return;
        }

        this.element.parentNode.removeChild(this.element);
        this.element = null;
        this.isShown = false;
        window.dispatchEvent(new CustomEvent('gaia-confirm-close'));
      });

      this.element.setAttribute('hidden', '');
    },

    // Scroll to up/down
    scrollContext(key) {
      const dialogContainer = this.parent.querySelector('div.content');
      const headerHeight = this.parent.querySelector('h1').clientHeight;
      const containerHeight =
        document.body.clientHeight -
        this.getSoftkeyHeight() -
        this.getStatusBarHeight() -
        headerHeight;
      dialogContainer.style.maxHeight = `${containerHeight}px`;
      // If dialog shows all texts , no need to scroll it
      if (dialogContainer.clientHeight < containerHeight) {
        this.scroll = false;
        return;
      }

      dialogContainer.scrollTop =
        key === 'ArrowUp'
          ? dialogContainer.scrollTop - this.deltaTop
          : dialogContainer.scrollTop + this.deltaTop;
    },

    getSoftkeyHeight() {
      let height = 0;
      const softKeyPanel = document.querySelector(`#${FORM_ID}`);
      if (softKeyPanel && softKeyPanel.classList.contains('visible')) {
        height = softKeyPanel.clientHeight;
      }
      return height;
    },

    getStatusBarHeight() {
      const style = getComputedStyle(document.body);
      const fontSize = parseFloat(style.fontSize);
      // 2 is 2rem from visual spec for the margin between statusbar and dialog.
      const rem = parseFloat(style.getPropertyValue('--statusbar-height')) + 2;
      return fontSize * rem;
    },

    handleEvent(e) {
      // Ensure we hide the dialog in the face of other errors.

      switch (e.type) {
        case 'hashchange':
          // Hashchange is only here to trigger this to call destroy.
          break;
        case 'confirm': {
          const confirm = this.config.confirm.callback;
          let cValue = null;
          if (this.input) {
            cValue = this.input.value.trim();
          }
          this.destroy();
          confirm && confirm(cValue);
          break;
        }
        case 'cancel': {
          this.destroy();
          const cancel = this.config.cancel.callback;
          cancel && cancel();
          break;
        }
        case 'accept': {
          this.destroy();
          const accept = this.config.accept.callback;
          accept && accept();
          break;
        }
        default:
          break;
      }
      // Handle back key
      switch (e.key) {
        case 'BrowserBack':
        case 'KanjiMode':
        case 'Backspace': {
          confirmSoftkey.hide();
          this.destroy();
          const back = this.config.backcallback;
          back && back();
          e.preventDefault();
          e.stopPropagation();
          break;
        }
        case 'Enter':
        case 'Accept':
          if (this.config.accept) {
            confirmSoftkey.hide();
            this.element.dispatchEvent(new CustomEvent('accept'));
          }
          break;
        case 'ContextMenu':
        case 'SoftLeft':
        case 'F1':
          if (this.config.cancel && this.isShown) {
            confirmSoftkey.hide();
            this.element.dispatchEvent(new CustomEvent('cancel'));
          }
          break;
        case 'ArrowUp':
        case 'ArrowDown':
          this.scrollContext(e.key);
          e.preventDefault();
          e.stopPropagation();
          break;
        case 'BrowserSearch':
        case 'SoftRight':
        case 'F2':
          if (this.input && this.input.value.trim().length < 1) {
            break;
          }
          if (this.config.confirm && this.isShown) {
            confirmSoftkey.hide();
            this.element.dispatchEvent(new CustomEvent('confirm'));
          }
          break;
        default:
          break;
      }
    }
  };

  // This name is intentionally verbose.
  window.ConfirmDialogHelper = ConfirmDialogHelper;
})(window);
