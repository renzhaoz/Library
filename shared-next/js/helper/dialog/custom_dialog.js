/* exported CustomDialog */
// XXXX: Waiting for the window.showModalDialog support in B2G
/* eslint-disable no-undef */

const CustomDialog = (function customDialog() {
  let container = null;
  let screen = null;
  let dialog = null;
  let header = null;
  let message = null;
  let input = null;
  let yes = null;
  let no = null;
  let backkeyCallback = null;
  let visibility = false;
  const callbacks = {
    cancel: null,
    confirm: null,
    accept: null,
    run: function run(type) {
      // eslint-disable-next-line no-unused-expressions
      this[type] && this[type].callback();
    }
  };
  const initSKs = function initSKs(actions) {
    if (CustomDialog.__version === 1) {
      if (typeof SoftkeyPanel === 'undefined') {
        LazyLoader.load(
          !window.SoftkeyPanel
            ? '%SHARED_APP_ORIGIN%/js/helper/softkey/softkey_panel.js'
            : [],
          function lazyLoaderCallback() {
            SoftkeyHelper.init(actions, function initCallback() {
              if (backkeyCallback !== null) {
                SoftkeyHelper.hide();
                backkeyCallback();
              }
            });
          }
        );
      } else {
        SoftkeyHelper.init(actions, function initCallback() {
          if (backkeyCallback !== null) {
            SoftkeyHelper.hide();
            backkeyCallback();
          }
        });
      }
    } else if (CustomDialog.__version === 2) {
      if (OptionHelper.softkeyPanel && OptionHelper.softkeyPanel.menuVisible) {
        OptionHelper.softkeyPanel.hideMenu();
      }
      OptionHelper.saveContext();
      OptionHelper.optionParams._for_custom_dialog = actions;
      OptionHelper.show('_for_custom_dialog');
    } else if (CustomDialog.__version === 3) {
      if (CustomDialog.softkeyPanel) {
        CustomDialog.softkeyPanel.initSoftKeyPanel(actions);
      }
    }
  };
  return {
    hide: function hide() {
      if (screen === null) {
        return;
      }

      if (!container) {
        container = document.body;
      }

      container.removeChild(screen);
      screen = null;
      dialog = null;
      header = null;
      message = null;
      yes = null;
      no = null;
      if (CustomDialog.__version === 2) {
        OptionHelper.returnContext();
      } else if (CustomDialog.__version === 1) {
        SoftkeyHelper.hide();
      }
      visibility = false;
      window.dispatchEvent(
        new CustomEvent('customDialogEvent', {
          detail: {
            visibility: false
          }
        })
      );
    },

    /**
     * Method that shows the dialog
     * @param  {Object} title the l10n id for the title of the dialog. null or
     *                  empty for no title. or you can give a object with more
     *                  options like {icon: path or empty string,
     *                  id: an l10n-id, args: an object of arguments that
     *                  parameterise the 'id' field}.
     * @param  {String} msg message for the dialog. give a object like the
     *                  title to enable more options.
     * @param  {Object} cancel {title, callback} object when confirm.
     * @param  {Object} confirm {title, callback} object when cancel.
     * @param  {Element} [containerElement] an optional element for which the
     *                   dialog will be injected
     */
    // eslint-disable-next-line max-params
    show: function show(
      title,
      msg,
      cancelAndBack,
      confirm,
      containerElement,
      center,
      inputType
    ) {
      const params = {
        menuClassName: 'menu-button',
        header: { l10nId: '' },
        items: []
      };
      container = containerElement || document.body;
      callbacks.confirm = confirm;
      callbacks.cancel = cancelAndBack;
      callbacks.accept = center;
      if (cancelAndBack) {
        if (cancelAndBack.backkeyCallback !== undefined) {
          ({ backkeyCallback } = cancelAndBack);
        }
        params.items.push({
          name: 'Cancel',
          l10nId: cancelAndBack.title,
          priority: 1,
          method() {
            if (CustomDialog.__version === 1) {
              SoftkeyHelper.hide();
            }
            cancelAndBack.callback();
          }
        });
      }
      if (center) {
        params.items.push({
          name: '',
          l10nId: center.title,
          priority: 2,
          method() {
            if (CustomDialog.__version === 1) {
              SoftkeyHelper.hide();
            }
            center.callback();
          }
        });
      }
      if (confirm) {
        params.items.push({
          name: 'Delete',
          l10nId: confirm.title,
          priority: 3,
          method() {
            if (CustomDialog.__version === 1) {
              SoftkeyHelper.hide();
            }
            if (input) {
              confirm.callback(input);
            } else {
              confirm.callback();
            }
          }
        });
      }
      initSKs(params);
      if (screen === null) {
        screen = document.createElement('form');
        screen.setAttribute('role', 'dialog');
        screen.setAttribute('data-type', 'confirm');
        screen.id = 'dialog-screen';

        dialog = document.createElement('section');
        screen.appendChild(dialog);

        /*
         * Create a reusable function to decorate elements with all
         * possible options, instead of scattering similar code about
         * everywhere.
         *
         * It's also possible to be extended with more usable decorating
         * options and elements.
         *
         * 'title'|'message' -> Object|String -> the element -> dialog
         * -> the decorated element
         */
        // eslint-disable-next-line max-params
        const decorateWithOptions = function decorateWithOptions(
          type,
          options,
          elm,
          // eslint-disable-next-line no-unused-vars,no-shadow
          dialog
        ) {
          if ('string' === typeof options) {
            elm.setAttribute('data-l10n-id', options);
            return elm;
          }

          const { icon } = options;

          let textElm = elm;

          if (icon && '' !== icon) {
            /*
             * We can't localize elements with child nodes
             * so if there's going to be an icon, we need to create a separate
             * node for text. See bug 1053629 for details
             */
            textElm = document.createElement('span');
            const iconImg = new Image();
            iconImg.src = icon;
            iconImg.classList.add(`custom-dialog-${type}-icon`);

            // Icons usually insert as the first element.
            elm.insertBefore(iconImg, elm.firstChild);
            elm.appendChild(textElm);
          }
          // More decorating options goes here.

          if (options.id) {
            window.api.l10n.setAttributes(textElm, options.id, options.args);
          } else {
            const text = options[type];
            textElm.textContent = text;
          }

          return elm;
        };

        const setElementText = function setElementText(element, options) {
          if ('string' === typeof options) {
            element.setAttribute('data-l10n-id', options);
          }

          if (options.id) {
            window.api.l10n.setAttributes(element, options.id, options.args);
          }
        };

        header = document.createElement('h1');
        header.id = 'dialog-title';
        if (title && title !== '') {
          header = decorateWithOptions('title', title, header, dialog);
        }
        dialog.appendChild(header);

        message = document.createElement('p');
        message.id = 'dialog-message';
        message = decorateWithOptions('message', msg, message, dialog);
        dialog.appendChild(message);

        if (inputType) {
          input = document.createElement('input');
          input.setAttribute('type', inputType);
          input.setAttribute('id', 'custom-input');
          confirm.input = input;
          dialog.appendChild(input);
        }

        if (cancelAndBack) {
          const menu = document.createElement('menu');
          menu.dataset.items = 1;

          no = document.createElement('button');

          /*
           * The default type of the button element is "Submit",
           * and form submit in system app would make system app reload.
           */
          no.type = 'button';

          setElementText(no, cancelAndBack.title);
          no.id = 'dialog-no';
          no.addEventListener('click', clickHandler);
          menu.appendChild(no);

          if (confirm) {
            menu.dataset.items = 2;
            yes = document.createElement('button');

            /*
             * The default type of button element is "Submit",
             * and form submit in system app would make system app reload.
             */
            yes.type = 'button';

            setElementText(yes, confirm.title);
            yes.id = 'dialog-yes';

            /*
             *Confirm can be with class "danger" or "recommend"
             *the default is "danger"
             */
            yes.className = confirm.recommend ? 'recommend' : 'danger';

            yes.addEventListener('click', clickHandler);
            menu.appendChild(yes);
          } else {
            // 1 button, should be full.
            no.classList.add('full');
          }
        }

        container.appendChild(screen);
      }

      // Make the screen visible
      screen.classList.add('visible');
      visibility = true;
      window.dispatchEvent(
        new CustomEvent('customDialogEvent', {
          detail: {
            visibility: true
          }
        })
      );

      // This is the event listener function for the buttons
      function clickHandler(evt) {
        // Hide the dialog
        screen.classList.remove('visible');

        // Call the appropriate callback, if it is defined
        if (evt.target === yes && confirm.callback) {
          confirm.callback();
        } else if (evt.target === no && cancel.callback) {
          cancel.callback();
        }
      }
      if (inputType) {
        input.focus();
      }
      return screen;
    },
    runConfirm() {
      callbacks.run('confirm');
    },
    runCancel() {
      callbacks.cancel ? callbacks.run('cancel') : callbacks.run('accept');
    },
    __version: 1,
    setVersion(version) {
      this.__version = version;
    },
    setSoftkeyPanel(softkeyPanel) {
      this.softkeyPanel = softkeyPanel;
    },
    get isVisible() {
      return visibility;
    }
  };
})();

window.CustomDialog = CustomDialog;
