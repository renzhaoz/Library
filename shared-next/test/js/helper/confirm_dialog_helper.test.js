/* eslint-disable no-undef, global-require */
describe('helper js <confirm_dialog_helper> test', () => {
  const { wrap } = require('jest-snapshot-serializer-raw');
  let dialog = null;

  beforeAll(done => {
    const {
      MockSoftkeyPanel,
      FORM_ID
    } = require('../../mock/softkeyPanel_mock');
    window.SoftkeyPanel = MockSoftkeyPanel;
    window.FORM_ID = FORM_ID;
    require('../../../js/helper/dialog/confirm_dialog_helper');

    const dialogConfig = {
      title: { id: 'error-pair-title', args: {} },
      body: { id: 'error-pair-fail', args: { devicename: 'deviceItem.name' } },
      desc: { id: 'error-pair-checkpin', args: {} },
      backcallback: jest.fn(),
      cancel: {
        l10nId: 'cancel',
        priority: 1,
        callback: jest.fn()
      },
      confirm: {
        l10nId: 'pair',
        priority: 3,
        callback: jest.fn()
      },
      accept: {
        l10nId: 'accept',
        priority: 3,
        callback: jest.fn()
      }
    };
    dialog = new ConfirmDialogHelper(dialogConfig);
    done();
  });

  describe('object/function existence test', () => {
    // Test ConfirmDialogHelper
    test('ConfirmDialogHelper should be function', done => {
      expect(typeof window.ConfirmDialogHelper).toBe('function');
      done();
    });

    // Test dialog.activeElement
    test('dialog.activeElement should be object', done => {
      expect(typeof dialog.activeElement).toBe('object');
      done();
    });

    // Test dialog.deltaTop
    test('dialog.deltaTop should be number', done => {
      expect(typeof dialog.deltaTop).toBe('number');
      done();
    });

    // Test dialog.Softkey
    test('dialog.Softkey should be object', done => {
      expect(typeof dialog.Softkey).toBe('object');
      done();
    });

    // Test dialog.isShown
    test('dialog.isShown should be boolean', done => {
      expect(typeof dialog.isShown).toBe('boolean');
      done();
    });

    // Test dialog.show
    test('dialog.show should be function', done => {
      expect(typeof dialog.show).toBe('function');
      done();
    });

    // Test dialog._show
    test('dialog._show should be function', done => {
      expect(typeof dialog._show).toBe('function');
      done();
    });

    // Test dialog.close
    test('dialog.close should be function', done => {
      expect(typeof dialog.close).toBe('function');
      done();
    });

    // Test dialog.destroy
    test('dialog.destroy should be function', done => {
      expect(typeof dialog.destroy).toBe('function');
      done();
    });

    // Test dialog.getSoftkeyHeight
    test('dialog.getSoftkeyHeight should be function', done => {
      expect(typeof dialog.getSoftkeyHeight).toBe('function');
      done();
    });

    // Test dialog.scrollContext
    test('dialog.scrollContext should be function', done => {
      expect(typeof dialog.scrollContext).toBe('function');
      done();
    });

    // Test dialog.handleEvent
    test('dialog.handleEvent should be function', done => {
      expect(typeof dialog.handleEvent).toBe('function');
      done();
    });
  });

  describe('function executed result test', () => {
    const parent = document.createElement('div');
    parent.id = 'confirm-dialog-container';
    beforeAll(done => {
      require('../../mock/lazyLoader_mock');
      require('../../mock/navigation_map');
      require('../../mock/l10n_mock');
      done();
    });

    // Test dialog.show
    test('dialog.show should invoke LazyLoader.load', done => {
      dialog.show();
      expect(LazyLoader.load).toBeCalledTimes(1);
      done();
    });

    // Test dialog.show
    test('dialog.show should appendchild to parent', done => {
      const openCallback = jest.fn();
      window.addEventListener('gaia-confirm-open', openCallback);
      LazyLoader.load.mockImplementationOnce((res, cb) => {
        cb();
      });
      jest.spyOn(document.body, 'clientHeight', 'get').mockReturnValue(960);
      document.body.setAttribute(
        'style',
        'font-size: 10px; --statusbar-height: 20px'
      );
      jest.spyOn(window, 'requestAnimationFrame').mockImplementation(cb => {
        dialog.isShown = true;
        cb();
      });
      dialog.show(parent);
      expect(LazyLoader.load).toBeCalledTimes(1);
      expect(dialog.Softkey.initSoftKeyPanel).toBeCalledTimes(1);
      expect(dialog.Softkey.show).toBeCalledTimes(1);
      expect(openCallback).toBeCalledTimes(1);
      expect(wrap(parent)).toMatchSnapshot();
      done();
    });

    // Test handle Event
    describe('handle Event test', () => {
      beforeEach(done => {
        LazyLoader.load.mockImplementationOnce((res, cb) => {
          cb();
        });
        window.NavigationMap = undefined;
        jest.spyOn(window, 'requestAnimationFrame').mockImplementation(cb => {
          cb();
        });
        dialog.show(parent);
        jest.resetAllMocks();
        done();
      });

      test('handle keydown BrowserBack', done => {
        window.dispatchEvent(
          new KeyboardEvent('keydown', { type: 'keydown', key: 'BrowserBack' })
        );
        expect(dialog.Softkey.hide).toBeCalledTimes(1);
        expect(dialog.config.backcallback).toBeCalledTimes(1);
        done();
      });

      test('handle keydown Enter', done => {
        window.dispatchEvent(
          new KeyboardEvent('keydown', { type: 'keydown', key: 'Enter' })
        );
        expect(dialog.Softkey.hide).toBeCalledTimes(1);
        done();
      });

      test('handle keydown ContextMenu', done => {
        window.dispatchEvent(
          new KeyboardEvent('keydown', { type: 'keydown', key: 'ContextMenu' })
        );
        expect(dialog.Softkey.hide).toBeCalledTimes(1);
        done();
      });

      test('handle keydown ArrowUp', done => {
        window.dispatchEvent(
          new KeyboardEvent('keydown', { type: 'keydown', key: 'ArrowUp' })
        );
        done();
      });

      test('handle keydown BrowserSearch', done => {
        window.dispatchEvent(
          new KeyboardEvent('keydown', {
            type: 'keydown',
            key: 'BrowserSearch'
          })
        );
        expect(dialog.Softkey.hide).toBeCalledTimes(1);
        done();
      });

      test('handle CustomEvent hashchange', done => {
        window.dispatchEvent(
          new CustomEvent('hashchange', { type: 'hashchange' })
        );
        done();
      });

      test('handle CustomEvent confirm/cancel/accept test', done => {
        dialog.element.dispatchEvent(
          new CustomEvent('confirm', { type: 'confirm' })
        );
        expect(dialog.config.confirm.callback).toBeCalledTimes(1);
        dialog.element.dispatchEvent(
          new CustomEvent('cancel', { type: 'cancel' })
        );
        expect(dialog.config.cancel.callback).toBeCalledTimes(1);
        dialog.element.dispatchEvent(
          new CustomEvent('accept', { type: 'accept' })
        );
        expect(dialog.config.accept.callback).toBeCalledTimes(1);
        done();
      });

      test('test dialog.close', done => {
        dialog.close();
        expect(dialog.Softkey.hide).toBeCalledTimes(1);
        done();
      });

      test('test dialog.destroy', done => {
        const closeCallback = jest.fn();
        window.addEventListener('gaia-confirm-close', closeCallback);
        jest.useFakeTimers();
        dialog.destroy();
        jest.runAllTimers();
        dialog.element.dispatchEvent(new CustomEvent('transitionend'));
        expect(dialog.element).toBeNull();
        expect(closeCallback).toBeCalledTimes(1);
        done();
      });

      afterEach(done => {
        jest.resetAllMocks();
        jest.restoreAllMocks();
        done();
      });
    });

    afterEach(done => {
      jest.resetAllMocks();
      jest.restoreAllMocks();
      done();
    });

    afterAll(done => {
      document.body.removeAttribute('style');
      done();
    });
  });

  afterAll(done => {
    window.SoftkeyPanel = undefined;
    window.FORM_ID = undefined;
    done();
  });
});
