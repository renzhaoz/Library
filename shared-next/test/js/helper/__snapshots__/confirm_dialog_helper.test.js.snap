// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helper js <confirm_dialog_helper> test function executed result test dialog.show should appendchild to parent 1`] = `
Object {
  Symbol(jest-snapshot-serializer-raw): <div
    id="confirm-dialog-container"
  >
    <gaia-confirm
      class=""
      data-type="undefined"
      role="menuitem"
      tabindex="-1"
    >
      <h1
        data-l10n-args="{}"
        data-l10n-id="error-pair-title"
        slot="title"
        style="text-align:center;"
      />
      <div
        class="content"
        slot="content"
        style="max-height: 740px;"
      >
        <div>
          <p
            data-l10n-args="{\\"devicename\\":\\"deviceItem.name\\"}"
            data-l10n-id="error-pair-fail"
            style="font-weight: normal;word-wrap: break-word;margin: 0;padding: 1rem;line-height: 2.4rem;color: var(--color-gs90);background: var(--color-gs10);"
          />
          <p
            class="noborder"
            data-l10n-args="{}"
            data-l10n-id="error-pair-checkpin"
            style="font-weight: normal;word-wrap: break-word;margin: 0;padding: 1rem;line-height: 2.4rem;color: var(--color-gs90);background: var(--color-gs10);"
          />
        </div>
      </div>
    </gaia-confirm>
  </div>,
}
`;
